subgraphs/compound-v3/src/sdk/account.ts

```diff
@@ -0,0 +1,331 @@
+    priceUSD: BigDecimal,
+    interestType: string | null = null
+  ): void {
+    let counterID = this.account.id
+      .toHexString()
+      .concat("-")
+      .concat(this.market.id.toHexString())
+      .concat("-")
+      .concat(side);
+    if (interestType) {
+      counterID = counterID.concat("-").concat(interestType);
+    }
+    const positionCounter = PositionCounter.load(counterID);
+    if (!positionCounter) {
+      log.warning("[subtractPosition] position counter {} not found", [
+        counterID,
+      ]);
+      this.positonError = true;
+      return;
+    }
+    const positionID = positionCounter.id
+      .concat("-")
+      .concat(positionCounter.nextCount.toString());
+    const position = Position.load(positionID);
+    if (!position) {
+      log.warning("[subtractPosition] position {} not found", [positionID]);
+      this.positonError = true;
+      return;
+    }
+
+    position.balance = newBalance;
+
+    if (transactionType == TransactionType.WITHDRAW) {
+      position.withdrawCount += INT_ONE;
+    } else if (transactionType == TransactionType.REPAY) {
+      position.repayCount += INT_ONE;
+    } else if (transactionType == TransactionType.TRANSFER) {
+      position.transferredCount += INT_ONE;
+    } else if (transactionType == TransactionType.LIQUIDATE) {
+      position.liquidationCount += INT_ONE;
+    }
+    position.save();
+
+    const closePosition = position.balance == BIGINT_ZERO;
+    if (closePosition) {
+      //
+      // update position counter
+      //
+      positionCounter.nextCount += INT_ONE;
+      positionCounter.save();
+
+      //
+      // close position
+      //
+      position.hashClosed = this.event.transaction.hash;
+      position.blockNumberClosed = this.event.block.number;
+      position.timestampClosed = this.event.block.timestamp;
+      position.save();
+
+      //
+      // update account position
+      //
+      this.account.openPositionCount -= INT_ONE;
+      this.account.closedPositionCount += INT_ONE;
+      this.account.save();
+
+      //
+      // update market position
+      //
+      this.market.openPositionCount -= INT_ONE;
+      this.market.closedPositionCount += INT_ONE;
+      this.market.save();
+
+      //
+      // update protocol position
+      //
+      this.protocol.openPositionCount -= INT_ONE;
+      this.protocol.save();
+    }
+    this.position = position;
+    this.positonError = false;
+
+    //
+    // update position snapshot
+    //
+    this.snapshotPosition(priceUSD);
+  }
+
+  countFlashloan(): void {
+    this.account.flashloanCount += INT_ONE;
+    this.account.save();
+  }
+
+  // Ensure this is called on the liquidators account
+  countLiquidate(): void {
+    this.account.liquidateCount += INT_ONE;
+    this.account.save();
+  }
+
+  getPositionID(): string | null {
+    if (!this.positonError) {
+      return this.position.id;
+    }
+    return null;
+  }
+
+  setCollateral(isCollateral: boolean): void {
```

Comment:
It seems it is possible for this function (and the ones below) to be called without `this.position` being set. You could prevent it by not exposing these directly from the AccountManager, but have it expose some `Account.getCurrentPositionForMarket(market: Market): PositionManager`. If the user can only modify positions via its own manager you ensure that a position is always initialized before attempting to modify it.

Reply:
True, and you may want to set something in the position even when you are not adding or subtracting to the position. Fixed in most recent commit